core/arch/arm/sm/sm_a32.S

```diff
@@ -249,6 +250,47 @@ END_FUNC sm_fiq_entry
 LOCAL_FUNC sm_vect_table , :
 UNWIND(	.fnstart)
 UNWIND(	.cantunwind)
+#ifdef CFG_CORE_WORKAROUND_SPECTRE_BP
+	/*
+	 * This depends on SP being 8 byte aligned, that is, the lowest
+	 * three bits in SP are zero.
+	 *
+	 * The idea is to form a specific bit pattern in the lowest three
+	 * bits of SP depending on which entry in the vector we enter via.
+	 * This is done by adding 1 to SP in each entry but the last.
+	 */
+	add	sp, sp, #1	/* 7:111 Reset			*/
+	add	sp, sp, #1	/* 6:110 Undefined instruction	*/
+	add	sp, sp, #1	/* 5:101 Secure monitor call	*/
+	add	sp, sp, #1	/* 4:100 Prefetch abort		*/
+	add	sp, sp, #1	/* 3:011 Data abort		*/
+	add	sp, sp, #1	/* 2:010 Reserved		*/
+	add	sp, sp, #1	/* 1:001 IRQ			*/
+	nop			/* 0:000 FIQ			*/
+
+	/* Invalidate the branch predictor for the current processor. */
+	write_bpiall
```

Comment:
Will this work on Cortex-A15?
https://github.com/OP-TEE/optee_os/pull/2055#issuecomment-356896233

Reply:
As far as I understand, not as intended at least.

Reply:
BPIALL is not effective in invalidating the branch predictor on A15. An ICALLU with the ACTLR[0] set to 1 need to be done on Cortex-A15. The [ARM TF advisory](https://github.com/ARM-software/arm-trusted-firmware/wiki/Arm-Trusted-Firmware-Security-Advisory-TFV-6) has been updated with this info. This would also mean the fix needs to detect the CPU type at runtime and apply the appropriate workaround.

Reply:
Addressed in https://github.com/OP-TEE/optee_os/pull/2065